tt_eager/tt_dnn/op_library/bmm/multi_core_reuse_mcast_2d_optimized/bmm_op_multi_core_reuse_mcast_2d_optimized.cpp

```diff
@@ -69,7 +71,24 @@ operation::ProgramWithCallbacks create_program_mcast_in0_in1(
     uint32_t out_CB_size = out_CB_tiles * output_single_tile_size;
     uint32_t interm0_CB_size = out_CB_tiles * interm0_single_tile_size;
 
-    uint32_t in2_block_tiles = per_core_M * in0_block_w;
+    uint32_t in2_block_tiles;
+    uint32_t in0_shard_width_in_tiles = 0;
+    uint32_t in0_shard_height_in_tiles = 0;
+    if (in0_is_sharded) {
+        in0_shard_width_in_tiles = in0_buffer->shard_spec().shape()[1] / 32;
+        in0_shard_height_in_tiles = in0_buffer->shard_spec().shape()[0] / 32;
+        in2_block_tiles = per_core_M * in0_shard_width_in_tiles;
+    } else {
+        in2_block_tiles = per_core_M * in0_block_w;
+    }
+
+    uint32_t in4_block_tiles = 0;
+    if (in0_is_sharded) {
+        if (in0_shard_height_in_tiles > 1 && in0_shard_width_in_tiles != in0_block_w) {
+            in4_block_tiles = per_core_M * in0_block_w;
+        }
+    }
```

Comment:
We should just mcast to in0 directly to avoid unncessary CBs, but if we were to do this with a separate CB, this needs to be double buffered.